// gwe/gfx2/gfx2_tilelayer.js
export class GfxTileLayer {
  private name: string;
  private grid: Array<number>;
  private rows: number;
  private columns: number;
  private visible: boolean;

  constructor() {
    this.name = '';
    this.grid = [];
    this.rows = 0;
    this.columns = 0;
    this.visible = true;
  }

  loadFromData(data: any): void {
    this.name = data['Name'];
    this.grid = data['Data'];
    this.rows = data['Rows'];
    this.columns = data['Columns'];
    this.visible = data['Visible'];
  }

  getLocationAt(index: number): Array<number> {
    const y = Math.floor(index / this.columns);
    const x = index % this.columns;
    return [x, y];
  }

  getTileCount(): number {
    return this.grid.length;
  }

  getTile(index: number): number {
    return this.grid[index];
  }

  getName(): string {
    return this.name;
  }

  getRows(): number {
    return this.rows;
  }

  getColumns(): number {
    return this.columns;
  }

  isVisible(): boolean {
    return this.visible;
  }

  setVisible(visible: boolean): void {
    this.visible = visible;
  }
}
