// gwe/gfx2/gfx2_map.js
let { gfx2TextureManager } = require('./gfx2_texture_manager');

class Gfx2Map {
  constructor() {
    this.rows = 0;
    this.columns = 0;
    this.tileHeight = 0;
    this.tileWidth = 0;
    this.tileLayers = [];
    this.tileset = new Gfx2Tileset();
  }

  async loadFromFile(path) {
    let response = await fetch(path);
    let json = await response.json();

    this.rows = json['Rows'];
    this.columns = json['Columns'];
    this.tileHeight = json['TileHeight'];
    this.tileWidth = json['TileWidth'];

    for (let obj of json['Layers']) {
      let tileLayer = new Gfx2TileLayer();
      await tileLayer.loadFromData(obj);
      this.tileLayers.push(tileLayer);
    }

    await this.tileset.loadFromData(json['Tileset']);
  }

  getWidth() {
    return this.columns * this.tileWidth;
  }

  getHeight() {
    return this.rows * this.tileHeight;
  }

  getRows() {
    return this.rows;
  }

  getColumns() {
    return this.columns;
  }

  getTileHeight() {
    return this.tileHeight;
  }

  getTileWidth() {
    return this.tileWidth;
  }

  getTileLayers() {
    return this.tileLayers;
  }

  getTileLayer(index) {
    return this.tileLayers[index];
  }

  getTileset() {
    return this.tileset;
  }
}

class Gfx2TileLayer {
  constructor() {
    this.name = '';
    this.grid = [];
    this.rows = 0;
    this.columns = 0;
    this.visible = true;
  }

  async loadFromData(data) {
    this.name = data['Name'];
    this.grid = data['Grid'];
    this.rows = data['Rows'];
    this.columns = data['Columns'];
    this.visible = data['Visible'];
  }

  getLocationAt(index) {
    let y = Math.floor(index / this.columns);
    let x = index % this.columns;
    return [x, y];
  }

  getTileCount() {
    return this.grid.length;
  }

  getTile(index) {
    return this.grid[index];
  }

  getName() {
    return this.name;
  }

  getRows() {
    return this.rows;
  }

  getColumns() {
    return this.columns;
  }

  isVisible() {
    return this.visible;
  }

  setVisible(visible) {
    this.visible = visible;
  }
}

class Gfx2Tileset {
  constructor() {
    this.textureWidth = 0;
    this.textureHeight = 0;
    this.tileWidth = 0;
    this.tileHeight = 0;
    this.columns = 0;
    this.texture = gfx2TextureManager.getDefaultTexture();
  }

  async loadFromData(data) {
    this.textureHeight = parseInt(data['TextureHeight']);
    this.textureWidth = parseInt(data['TextureWidth']);
    this.tileWidth = parseInt(data['TileWidth']);
    this.tileHeight = parseInt(data['TileHeight']);
    this.columns = parseInt(data['Columns']);
    this.texture = await gfx2TextureManager.loadTexture(data['TextureFile']);
  }

  getTilePosition(tileId) {
    let x = ((tileId - 1) % this.columns) * this.tileWidth;
    let y = Math.floor((tileId - 1) / this.columns) * this.tileHeight;
    return [x, y];
  }

  getTexture() {
    return this.texture;
  }

  getColumn() {
    return this.columns;
  }

  getTileHeight() {
    return this.tileHeight;
  }

  getTileWidth() {
    return this.tileWidth;
  }

  getTextureHeight() {
    return this.textureHeight;
  }

  getTextureWidth() {
    return this.textureWidth;
  }
}

module.exports.Gfx2Map = Gfx2Map;
